### PR TITLE
JSDoc Types definition

### PR DESCRIPTION
**Pull request by TheJaredWilcurt**

### CODE REVIEW

**Review by AngelCMHxD — CHANGES_REQUESTED** [review]

File: index.js [hunk 1]
```diff
@@ -7,17 +7,74 @@
...
+ * @property {string} [name]                 Name of the shortcut file.
+ * @property {string} [comment]              Metadata file "comment" property. Description of what the shortcut would open.
+ * @property {string} [icon]                 Image shown on the shortcut icon. You can also pass in an index if file contains multiple icons, like `'C:\\file.exe,0'`
+ * @property {string} [arguments]            Additional arguments passed in to the end of your target `filePath`.
+ * @property {string} [windowMode="normal"]  How the window should be displayed by default. Valid inputs: 'normal', 'maximized', 'minimized'. Defaults to 'normal'.
+ * @property {string} [hotkey]               A global hotkey to associate to opening this shortcut, like 'CTRL+ALT+F'.
+ * @property {string} [workingDirectory]     The working directory for the shortcut when it launches, must be a valid path to a folder.
+ * @property {string} [VBScriptPath]         This is an advanced option specifically and only for projects packaged with `pkg`. [See documentation](https://github.com/nwutils/create-desktop-shortcuts#windows-settings).
```
[the remark this answers is not in the document]
Comment (AngelCMHxD) [reply index.js:33]: I don't think there is a way to do that.

File: index.js [hunk 2]
```diff
@@ -8,16 +8,76 @@ const validation = require('./src/validation.js');
...
- * Creates OS based shortcuts for files, folders, and applications.
+ * OPTIONAL: console.error is called by default if verbose: true.
+ *
+ * Your own custom logging function called with helpful warning/error
+ * messages from the internal validators. Only used if verbose: true.
+ *
+ * @callback {Function} CUSTOMLOGGER
+ * @param    {string}   message       The human readable warning/error message
+ * @param    {object}   [error]       Sometimes an error or options object is passed
+ * @param    {Array}    [asdf]        asdf
```
Comment (AngelCMHxD) [inline index.js:19]: Did you wanted to push these lines?

Comment (TheJaredWilcurt) [reply index.js:19]: If you are asking about the formatting (number of spaces), yes, I added a rule to my linter to include `@callback` in the formatting, and then the linter fixed the spacing.

File: index.js [hunk 3]
```diff
@@ -8,16 +8,76 @@ const validation = require('./src/validation.js');
...
+ * @param    {boolean}  [qwer]        qwer
+ * @param    {number}   [wert]        wert
```
Comment (TheJaredWilcurt) [inline index.js:21]:
```suggestion
```

File: index.js (hunk 2 above)
Comment (TheJaredWilcurt) [reply index.js]: Ah, I see what you were referring to now. cleaned up

File: index.js [hunk 4]
```diff
@@ -8,16 +8,73 @@ const validation = require('./src/validation.js');
...
+ * @property {string} [name]                 Name of the shortcut file.
+ * @property {string} [comment]              Metadata file "comment" property. Description of what the shortcut would open.
+ * @property {string} [icon]                 Image shown on the shortcut icon. You can also pass in an index if file contains multiple icons, like `'C:\\file.exe,0'`
+ * @property {string} [arguments]            Additional arguments passed in to the end of your target `filePath`.
+ * @property {string} [windowMode="normal"]  How the window should be displayed by default. Valid inputs: 'normal', 'maximized', 'minimized'. Defaults to 'normal'.
+ * @property {string} [hotkey]               A global hotkey to associate to opening this shortcut, like 'CTRL+ALT+F'.
+ * @property {string} [workingDirectory]     The working directory for the shortcut when it launches, must be a valid path to a folder.
+ * @property {string} [VBScriptPath]         This is an advanced option specifically and only for projects packaged with `pkg`. [See documentation](https://github.com/nwutils/create-desktop-shortcuts#windows-settings).
```
Comment (AngelCMHxD) [inline index.js:33]:
```suggestion
 * @property {string} [VBScriptPath]         This is an advanced option specifically and only for projects packaged with `pkg`.
 *                                           [See documentation](https://github.com/nwutils/create-desktop-shortcuts#windows-settings).
```

File: index.js [hunk 5]
```diff
@@ -8,16 +8,73 @@ const validation = require('./src/validation.js');
...
+ * @property {boolean} [chmod=true]      If true, will apply a `chmod +x` (755) to the shortcut after creation to allow execution permission.
+ * @property {string}  [arguments]       Additional arguments passed in to the end of your target `filePath`.
+ */
+
+/**
+ * @typedef  {object}  OSX
+ * @property {string}  filePath           The target the shortcut points to.
+ * @property {string}  [outputPath]       Path where shortcut will be placed. Defaults to user's desktop.
+ * @property {string}  [name]             Name of the shortcut file.
+ * @property {boolean} [overwrite=false]  If true, will replace any existing file in the `outputPath` with matching `name`. [See documentation](https://github.com/nwutils/create-desktop-shortcuts#osx-settings).
```
Comment (AngelCMHxD) [inline index.js:54]:
```suggestion
 * @property {boolean} [overwrite=false]  If true, will replace any existing file in the `outputPath` with matching `name`.
 *                                        [See documentation](https://github.com/nwutils/create-desktop-shortcuts#osx-settings).
```

**Review by AngelCMHxD — CHANGES_REQUESTED** [review]
After some testing I found out that splitting the line actually works out as there isn't another tag that makes the new line separate from the previous one and the resulting the pop-up is the same as before.

**Comment by TheJaredWilcurt** [comment]
That's really nice. I was looking into moving the API JSDoc definitions out to their own file, to reuse the OPTIONS definition in multiple places, just to see if that was possible. I was just trying to get the VSCode hover to work, but couldn't find a good resource. Seems that importing a JSDoc definition from a separate file is done differently with different tools and compatibility varies. I think it might be possible, but I don't really have time at the moment to look into it, it's unfortunately not a well documented feature.

**Comment by AngelCMHxD** [comment]
Is this what you wanted to do?

If you want to export another type def, you just need to do the same thing, for example, exporting osx options type:
```js
/**@type {OSX} */
let osxType;

module.exports = { optionsType, /*...*/ osxType }
```

Also, if you want to add more types def to that file, I recommend renaming the `OSX`, `LINUX` and `WINDOWS` types as that names could possibly conflict with what you want to do.

**Comment by TheJaredWilcurt** [comment]
That is awesome, I'm actually very happy with this result. Going to merge and do a new release.

**Pull request merged**